### ISSUE TITLE
Plex not getting updated on latest version of this cookbook?

### ISSUE THREAD

**Issue by JHBoricua** [meringu/plex#5]
For some reason the latest version of the cookbook is not updating plex. I'm running this cookbook on an Ubuntu 16.04 server. Currently the version of Plex I have is 1.2.3.2914-1ff0f18. When I run the chef-client, I see that the plex cookbook  downloads the latest version (currently 1.2.7.2987-1bef33a) fine. However when it goes to install it simply sees the existing one as being up-to-date so nothing is done.

Here's the output of my chef-client run:

```
linuxadmin@plex01:~$ sudo chef-client
Starting Chef Client, version 12.16.42
resolving cookbooks for run list: ["chef-client", "plex"]
Synchronizing Cookbooks:
  - cron (3.0.0)
  - chef-client (7.0.0)
  - logrotate (2.1.0)
  - compat_resource (12.16.2)
  - windows (2.0.2)
  - plex (0.2.2)
Installing Cookbook Gems:
Compiling Cookbooks...
Converging 9 resources
Recipe: chef-client::systemd_service
  * directory[/var/run/chef] action create (up to date)
  * directory[/var/cache/chef] action create (up to date)
  * directory[/var/lib/chef] action create (up to date)
  * directory[/var/log/chef] action create (up to date)
  * directory[/etc/chef] action create (up to date)
  * template[/etc/systemd/system/chef-client.service] action create (up to date)
  * template[/etc/default/chef-client] action create (up to date)
  * service[chef-client] action enable (up to date)
  * service[chef-client] action start (up to date)
Recipe: plex::default
  * plex_media_server[pms] action install
    * remote_file[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] action create (up to date)
    * dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] action install (up to date)
     (up to date)
  * plex_media_server[pms] action start
    * service[plexmediaserver] action start (up to date)
     (up to date)
  * plex_media_server[pms] action enable
    * service[plexmediaserver] action enable (up to date)
     (up to date)

Running handlers:
Running handlers complete
Chef Client finished, 0/16 resources updated in 04 seconds


linuxadmin@plex01:~$ dpkg -s plexmediaserver
Package: plexmediaserver
Status: install ok installed
Priority: extra
Section: video
Installed-Size: 214892
Maintainer: Plex Inc <support@plex.tv>
Architecture: amd64
Version: 1.2.3.2914-1ff0f18
Conffiles:
 /etc/default/plexmediaserver e70c95f4c5a72536f038437bedfcb2d6
 /etc/apt/sources.list.d/plexmediaserver.list 502be61f579d469a0cc962ee8188482f
 /etc/init/plexmediaserver.conf 2d1e30d693137f96d04da91cd34e8fe7
Description: Plex organizes all of your personal media so you can easily access and enjoy it.
Homepage: https://plex.tv
```

**Comment by JHBoricua** [comment]
I ran the chef-client with the debug flag and saw that the cookbook is not upgrading the existing chef package if a previous version exists. Not sure if this is the intended behavior, as in the past it would upgrade the existing package to the latest version. Here's the output:

```
[2016-11-18T12:16:22-06:00] INFO: Processing dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] action install (/var/chef/cache/cookbooks/plex/providers/media_server.rb line 40)
[2016-11-18T12:16:22-06:00] DEBUG: Providers for generic dpkg_package resource enabled on node include: [Chef::Provider::Package::Dpkg]
[2016-11-18T12:16:22-06:00] DEBUG: Provider for action install on resource dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] is Chef::Provider::Package::Dpkg
[2016-11-18T12:16:22-06:00] DEBUG: No package source specified, but /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb exists on filesystem, using /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb as source.
[2016-11-18T12:16:22-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] checking /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb dpkg status
[2016-11-18T12:16:22-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] checking install state of plexmediaserver
[2016-11-18T12:16:23-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] current version is 1.2.3.2914-1ff0f18
[2016-11-18T12:16:23-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb 1.2.3.2914-1ff0f18 already installed
[2016-11-18T12:16:23-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb] is already installed - nothing to do

```

**Comment by meringu** [comment]
This line 

```
/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb 1.2.3.2914-1ff0f18 already installed
```

would indicate the dpkg_package provider thinks the version of the `/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb` package is `1.2.3.2914-1ff0f18`.

I downloaded the package on an Ubuntu 16.04 VM to attempt to reproduce this.

```
$ dpkg-deb -W ~/Downloads/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb
plexmediaserver    1.2.7.2987-1bef33a
```

As you can see the copy I had shows the correct version.

Can you please check the output of 

```
dpkg-deb -W /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb
```

I suspect you will get the wrong version. If so, in this order can you please:
- delete `/var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb`.
- run chef-client, paste the debug log here.
- paste here the output of `dpkg -s plexmediaserver`
- paste here the output of `dpkg-deb -W /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb`.

Given the result, we can work out if any change is required in this cookbook.

**Comment by JHBoricua** [comment]
Hello,

The plot thickens. The output of 

`dpkg-deb -W /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb`

is actually reporting the correct version.

```
linuxadmin@plex01:~$ dpkg-deb -W /var/chef/cache/plexmediaserver_1.2.7.2987-1bef33a_amd64.deb
plexmediaserver 1.2.7.2987-1bef33a
```

Do you still want me to do the other steps?

**Comment by JHBoricua** [comment]
So this might have been a false alarm. Previously I was running plex on a centos server. I had initially installed plex on it manually but then transitioned to your cookbook once I started playing with chef on my home server lab. The cookbook took over the manual process I was doing to keep plex up to date.

Recently I rebuilt my entire server lab, moving away from ESXi hosts and to Proxmox VE. I then settled on Ubuntu server as my platform of choice and spun up a new Ubuntu VM for Plex. Since I didn't have chef fully deployed to my new server lab at the time, **I manually installed Plex by creating a custom apt repository in /etc/apt/sources.list.d/ and installing via apt**, rather than downloading the Plex .deb file manually and using dpkg to install it.

Ultimately I think this is what caused the issue when I then transitioned to chef and started using your recipe again. So I uninstalled plex via apt, removed the custom plex apt repo from /etc/sources.list.d and then ran chef client to have it install plex via the dpkg provider it uses. My guess is that when the next version of plex is released, the recipe will be able to upgrade it just fine.

**Comment by JHBoricua** [comment]
Ok, it is definitely not updating it when chef-client runs. When I closed this I had removed the plex apt repo I had on this server and wiped plex from it, which allowed the installation of plex via the chef-client recipe. The version installed back then (22 days ago) when I thought I had found the conflict was 1.2.7.2987-1bef33a. 

As of today Plex is at version 1.3.2.3112-1751929. The behavior is the same as reported originally. The recipe downloads the new package but when it compares the new version with the existing one installed on the server it erroneously determines that it is up to date and doesn't perform the update.

Here's the relevant section of the chef-client debug log, right after it's done downloading the new package and goes into the install routine:

```
    * dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] action install[2016-12-10T21:05:48-06:00] INFO: Processing dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] action install (/var/chef/cache/cookbooks/plex/providers/media_server.rb line 40)
[2016-12-10T21:05:48-06:00] DEBUG: Providers for generic dpkg_package resource enabled on node include: [Chef::Provider::Package::Dpkg]
[2016-12-10T21:05:48-06:00] DEBUG: Provider for action install on resource dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] is Chef::Provider::Package::Dpkg
[2016-12-10T21:05:48-06:00] DEBUG: No package source specified, but /var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb exists on filesystem, using /var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb as source.
[2016-12-10T21:05:48-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] checking /var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb dpkg status
plexmediaserver 1.3.2.3112-1751929
[2016-12-10T21:05:48-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] checking install state of plexmediaserver
Package: plexmediaserver
Status: install ok installed
Priority: extra
Section: video
Installed-Size: 215052
Maintainer: Plex Inc <support@plex.tv>
Architecture: amd64
Version: 1.2.7.2987-1bef33a
Conffiles:
 /etc/default/plexmediaserver e70c95f4c5a72536f038437bedfcb2d6
 /etc/apt/sources.list.d/plexmediaserver.list 502be61f579d469a0cc962ee8188482f
 /etc/init/plexmediaserver.conf 2d1e30d693137f96d04da91cd34e8fe7
Description: Plex organizes all of your personal media so you can easily access and enjoy it.
Homepage: https://plex.tv
[2016-12-10T21:05:48-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] current version is 1.2.7.2987-1bef33a
[2016-12-10T21:05:48-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] /var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb 1.2.7.2987-1bef33a already installed
[2016-12-10T21:05:48-06:00] DEBUG: dpkg_package[/var/chef/cache/plexmediaserver_1.3.2.3112-1751929_amd64.deb] is already installed - nothing to do
 (up to date)
```
Notice that when it checks the install state of the existing plex software on my server the version reported back is older than that of the new deb package that was downloaded, but for some reason the process thinks that they are the same version and thus it won't upgrade it.

**Comment by JHBoricua** [comment]
So I got this working by changing the code on the plex/providers/media_server.rb file in my local copy of your recipe. I changed lines 35 & 36, which uses the dpkg_package resource:
```
when 'debian'
    dpkg_package installer_file
```
to this, which uses the package resource with the upgrade action but calls the dpkg provider:
```
when 'debian'
    package 'plexmediaserver' do
      action :upgrade
      source installer_file
      provider Chef::Provider::Package::Dpkg
    end
```
Here's what I found interesting. When I had the action as 'install' on that new code, it would behave the same as I described before. When I changed the action to 'upgrade', it then upgraded the package. It seems the default 'install' action of the dpkg_package resource won't overwrite/upgrade a package that's already installed, regardless of the version difference. So using the package resource instead gets around that limitation. The 'upgrade' action also installs the package if its not present.

I'm a total newbie at this so feel free to tell me if I'm on the wrong track here.

**Comment by meringu** [comment]
I have reproduce this, chef debug log reports the different versions of the package installed as desired then chooses not to install it. I have merged your PR and released version 0.2.3.

I'll close this issue, please open it again if you are still experiencing a problem.
